File: web-app/server/routes/exports.js
/**
 * Data Export Routes Module
 * Handles CSV/JSON export for various data types
 */

const express = require('express');
const router = express.Router();
const { getDB } = require('../db');
const logger = require('../logger');

/**
 * Helper: Convert array of objects to CSV string
 */
function toCSV(data, columns) {
  if (!data || data.length === 0) return '';
  
  const headers = columns.map(c => c.label).join(',');
  const rows = data.map(row => {
    return columns.map(c => {
      let val = c.getter ? c.getter(row) : (row[c.key] ?? '');
      // Escape CSV special characters
      val = String(val).replace(/"/g, '""');
      if (val.includes(',') || val.includes('"') || val.includes('\n')) {
        val = `"${val}"`;
      }
      return val;
    }).join(',');
  });
  
  return [headers, ...rows].join('\n');
}

/**
 * GET /api/exports/products
 * Export products as CSV
 */
router.get('/products', async (req, res) => {
  try {
    const db = getDB();
    const products = await db.collection('products').find({}).sort({ name: 1 }).toArray();
    
    const columns = [
      { key: 'name', label: 'Product Name' },
      { key: 'barcode', label: 'Barcode' },
      { key: 'hsnCode', label: 'HSN Code' },
      { key: 'costPrice', label: 'Cost Price' },
      { key: 'sellingPrice', label: 'Selling Price' },
      { key: 'quantity', label: 'Stock' },
      { key: 'minStock', label: 'Min Stock' },
      { key: 'category', label: 'Category' },
      { label: 'Profit Margin', getter: (p) => p.sellingPrice && p.costPrice ? ((p.sellingPrice - p.costPrice) / p.costPrice * 100).toFixed(1) + '%' : 'N/A' }
    ];
    
    const csv = toCSV(products, columns);
    
    res.setHeader('Content-Type', 'text/csv');
    res.setHeader('Content-Disposition', 'attachment; filename=products_export.csv');
    res.send(csv);
  } catch (e) {
    logger.error('Product export error:', e);
    res.status(500).json({ error: e.message });
  }
});

/**
 * GET /api/exports/customers
 * Export customers as CSV
 */
router.get('/customers', async (req, res) => {
  try {
    const db = getDB();
    const customers = await db.collection('customers').find({}).sort({ name: 1 }).toArray();
    
    const columns = [
      { key: 'name', label: 'Customer Name' },
      { key: 'phone', label: 'Phone' },
      { key: 'email', label: 'Email' },
      { key: 'address', label: 'Address' },
      { label: 'Total Purchases', getter: (c) => c.totalPurchases || 0 },
      { label: 'Total Spent', getter: (c) => c.totalSpent || 0 },
      { label: 'Created Date', getter: (c) => c.createdAt ? new Date(c.createdAt).toLocaleDateString() : 'N/A' }
    ];
    
    const csv = toCSV(customers, columns);
    
    res.setHeader('Content-Type', 'text/csv');
    res.setHeader('Content-Disposition', 'attachment; filename=customers_export.csv');
    res.send(csv);
  } catch (e) {
    logger.error('Customer export error:', e);
    res.status(500).json({ error: e.message });
  }
});

/**
 * GET /api/exports/invoices
 * Export invoices as CSV with optional date range
 */
router.get('/invoices', async (req, res) => {
  try {
    const db = getDB();
    const { startDate, endDate } = req.query;
    
    const filter = {};
    if (startDate || endDate) {
      filter.billDate = {};
      if (startDate) filter.billDate.$gte = new Date(startDate);
      if (endDate) filter.billDate.$lte = new Date(endDate);
    }
    
    const invoices = await db.collection('bills').find(filter).sort({ billDate: -1 }).toArray();
    
    const columns = [
      { label: 'Invoice #', getter: (i) => i.billNumber || i._id.toString() },
      { label: 'Date', getter: (i) => i.billDate ? new Date(i.billDate).toLocaleDateString() : 'N/A' },
      { label: 'Customer', getter: (i) => i.customerName || 'Walk-in' },
      { label: 'Items', getter: (i) => (i.items || []).length },
      { label: 'Subtotal', getter: (i) => i.subtotal || 0 },
      { label: 'Discount', getter: (i) => i.discountAmount || 0 },
      { label: 'GST', getter: (i) => i.gstAmount || 0 },
      { label: 'Grand Total', getter: (i) => i.grandTotal || i.total || 0 },
      { label: 'Payment Mode', getter: (i) => i.paymentMode || 'N/A' },
      { label: 'Status', getter: (i) => i.paymentStatus || 'N/A' },
      { label: 'Cashier', getter: (i) => i.username || 'N/A' }
    ];
    
    const csv = toCSV(invoices, columns);
    
    res.setHeader('Content-Type', 'text/csv');
    res.setHeader('Content-Disposition', 'attachment; filename=invoices_export.csv');
    res.send(csv);
  } catch (e) {
    logger.error('Invoice export error:', e);
    res.status(500).json({ error: e.message });
  }
});

/**
 * GET /api/exports/expenses
 * Export expenses as CSV
 */
router.get('/expenses', async (req, res) => {
  try {
    const db = getDB();
    const { startDate, endDate } = req.query;
    
    const filter = {};
    if (startDate || endDate) {
      filter.date = {};
      if (startDate) filter.date.$gte = new Date(startDate);
      if (endDate) filter.date.$lte = new Date(endDate);
    }
    
    const expenses = await db.collection('expenses').find(filter).sort({ date: -1 }).toArray();
    
    const columns = [
      { key: 'description', label: 'Description' },
      { key: 'amount', label: 'Amount' },
      { key: 'category', label: 'Category' },
      { label: 'Date', getter: (e) => e.date ? new Date(e.date).toLocaleDateString() : 'N/A' },
      { key: 'createdByUsername', label: 'Created By' }
    ];
    
    const csv = toCSV(expenses, columns);
    
    res.setHeader('Content-Type', 'text/csv');
    res.setHeader('Content-Disposition', 'attachment; filename=expenses_export.csv');
    res.send(csv);
  } catch (e) {
    logger.error('Expense export error:', e);
    res.status(500).json({ error: e.message });
  }
});

/**
 * GET /api/exports/returns
 * Export returns as CSV
 */
router.get('/returns', async (req, res) => {
  try {
    const db = getDB();
    const returns = await db.collection('returns').find({}).sort({ createdAt: -1 }).toArray();
    
    const columns = [
      { key: 'billNumber', label: 'Invoice #' },
      { key: 'customerName', label: 'Customer' },
      { label: 'Items Returned', getter: (r) => (r.items || []).map(i => `${i.name} x${i.quantity}`).join('; ') },
      { key: 'refundAmount', label: 'Refund Amount' },
      { key: 'reason', label: 'Reason' },
      { key: 'status', label: 'Status' },
      { label: 'Date', getter: (r) => r.createdAt ? new Date(r.createdAt).toLocaleDateString() : 'N/A' },
      { key: 'processedByUsername', label: 'Processed By' }
    ];
    
    const csv = toCSV(returns, columns);
    
    res.setHeader('Content-Type', 'text/csv');
    res.setHeader('Content-Disposition', 'attachment; filename=returns_export.csv');
    res.send(csv);
  } catch (e) {
    logger.error('Returns export error:', e);
    res.status(500).json({ error: e.message });
  }
});

module.exports = router;

File: server-flask/routes/exports.py
import csv
import io
import logging
from datetime import datetime
from flask import Blueprint, request, send_file, jsonify
from database import get_db
from utils.auth_middleware import authenticate_token

logger = logging.getLogger(__name__)

exports_bp = Blueprint('exports', __name__)

@exports_bp.route('/products', methods=['GET'])
@authenticate_token
def export_products():
    """Export all products to CSV"""
    db = get_db()
    products = list(db.products.find().sort("name", 1))
    
    output = io.StringIO()
    writer = csv.writer(output)
    
    # Header row
    writer.writerow(['Product Name', 'Category', 'Price (Inclusive)', 'Cost Price', 'Stock Quantity', 'HSN/SKU', 'Status'])
    
    for p in products:
        writer.writerow([
            p.get('name', 'N/A'),
            p.get('category', 'General'),
            p.get('price', 0),
            p.get('costPrice', 0),
            p.get('quantity', 0),
            p.get('hsnCode', 'N/A'),
            'In Stock' if p.get('quantity', 0) > 0 else 'Out of Stock'
        ])
    
    output.seek(0)
    return send_file(
        io.BytesIO(output.getvalue().encode('utf-8')),
        mimetype='text/csv',
        as_attachment=True,
        download_name=f"products_export_{datetime.now().strftime('%Y%m%d')}.csv"
    )

@exports_bp.route('/customers', methods=['GET'])
@authenticate_token
def export_customers():
    """Export all customers to CSV"""
    db = get_db()
    customers = list(db.customers.find().sort("name", 1))
    
    output = io.StringIO()
    writer = csv.writer(output)
    
    # Header row
    writer.writerow(['Name', 'Phone', 'Email', 'Address', 'Place', 'City', 'GSTIN', 'Member Since'])
    
    for c in customers:
        created_at = c.get('createdAt')
        date_str = created_at.strftime('%Y-%m-%d') if isinstance(created_at, datetime) else 'N/A'
        
        writer.writerow([
            c.get('name', 'N/A'),
            c.get('phone', 'N/A'),
            c.get('email', 'N/A'),
            c.get('address', 'N/A'),
            c.get('place', 'N/A'),
            c.get('city', 'N/A'),
            c.get('gstin', 'N/A'),
            date_str
        ])
    
    output.seek(0)
    return send_file(
        io.BytesIO(output.getvalue().encode('utf-8')),
        mimetype='text/csv',
        as_attachment=True,
        download_name=f"customers_export_{datetime.now().strftime('%Y%m%d')}.csv"
    )

@exports_bp.route('/invoices', methods=['GET'])
@authenticate_token
def export_invoices():
    """Export invoice history to CSV with date range filtering"""
    db = get_db()
    start_date = request.args.get('startDate')
    end_date = request.args.get('endDate')
    
    query = {}
    if start_date or end_date:
        query['billDate'] = {}
        if start_date:
            query['billDate']['$gte'] = datetime.strptime(start_date, '%Y-%m-%d')
        if end_date:
            # End of day for the end_date
            query['billDate']['$lte'] = datetime.strptime(end_date, '%Y-%m-%d').replace(hour=23, minute=59, second=59)
            
    bills = list(db.bills.find(query).sort("billDate", -1))
    
    output = io.StringIO()
    writer = csv.writer(output)
    
    # Header row
    writer.writerow(['Invoice No', 'Date', 'Customer Name', 'Subtotal', 'Tax (GST)', 'Grand Total', 'Profit', 'Payment Mode'])
    
    for b in bills:
        bill_date = b.get('billDate')
        date_str = bill_date.strftime('%Y-%m-%d %H:%M') if isinstance(bill_date, datetime) else str(bill_date)
        
        writer.writerow([
            b.get('billNumber', 'N/A'),
            date_str,
            b.get('customerName', 'Walk-in'),
            round(b.get('subtotal', 0), 2),
            round(b.get('gstAmount', 0), 2),
            round(b.get('grandTotal', 0), 2),
            round(b.get('totalProfit', 0), 2),
            b.get('paymentMode', 'cash')
        ])
    
    output.seek(0)
    return send_file(
        io.BytesIO(output.getvalue().encode('utf-8')),
        mimetype='text/csv',
        as_attachment=True,
        download_name=f"invoices_export_{datetime.now().strftime('%Y%m%d')}.csv"
    )

@exports_bp.route('/expenses', methods=['GET'])
@authenticate_token
def export_expenses():
    """Export business expenses to CSV"""
    db = get_db()
    start_date = request.args.get('startDate')
    end_date = request.args.get('endDate')
    
    query = {}
    if start_date or end_date:
        query['date'] = {}
        if start_date:
            query['date']['$gte'] = datetime.strptime(start_date, '%Y-%m-%d')
        if end_date:
            query['date']['$lte'] = datetime.strptime(end_date, '%Y-%m-%d').replace(hour=23, minute=59, second=59)
            
    expenses = list(db.expenses.find(query).sort("date", -1))
    
    output = io.StringIO()
    writer = csv.writer(output)
    
    # Header row
    writer.writerow(['Date', 'Description', 'Category', 'Amount', 'Mode'])
    
    for e in expenses:
        exp_date = e.get('date')
        date_str = exp_date.strftime('%Y-%m-%d') if isinstance(exp_date, datetime) else str(exp_date)
        
        writer.writerow([
            date_str,
            e.get('description', 'N/A'),
            e.get('category', 'General'),
            round(e.get('amount', 0), 2),
            e.get('mode', 'cash')
        ])
    
    output.seek(0)
    return send_file(
        io.BytesIO(output.getvalue().encode('utf-8')),
        mimetype='text/csv',
        as_attachment=True,
        download_name=f"expenses_export_{datetime.now().strftime('%Y%m%d')}.csv"
    )

@exports_bp.route('/returns', methods=['GET'])
@authenticate_token
def export_returns():
    """Export product returns to CSV"""
    db = get_db()
    start_date = request.args.get('startDate')
    end_date = request.args.get('endDate')
    
    query = {}
    if start_date or end_date:
        query['returnDate'] = {}
        if start_date:
            query['returnDate']['$gte'] = datetime.strptime(start_date, '%Y-%m-%d')
        if end_date:
            query['returnDate']['$lte'] = datetime.strptime(end_date, '%Y-%m-%d').replace(hour=23, minute=59, second=59)
            
    returns = list(db.returns.find(query).sort("returnDate", -1))
    
    output = io.StringIO()
    writer = csv.writer(output)
    
    # Header row
    writer.writerow(['Return ID', 'Date', 'Invoice No', 'Product Name', 'Quantity', 'Refund Amount', 'Reason'])
    
    for r in returns:
        return_date = r.get('returnDate')
        date_str = return_date.strftime('%Y-%m-%d') if isinstance(return_date, datetime) else str(return_date)
        
        writer.writerow([
            str(r.get('_id')),
            date_str,
            r.get('billNumber', 'N/A'),
            r.get('productName', 'N/A'),
            r.get('quantity', 0),
            round(r.get('refundAmount', 0), 2),
            r.get('reason', 'N/A')
        ])
    
    output.seek(0)
    return send_file(
        io.BytesIO(output.getvalue().encode('utf-8')),
        mimetype='text/csv',
        as_attachment=True,
        download_name=f"returns_export_{datetime.now().strftime('%Y%m%d')}.csv"
    )
